### PR TITLE
refactor: with memoizeAtom

### PR DESCRIPTION
**Pull request by dai-shi**
putting typical pattern in a function. better typing.

### CODE REVIEW

File: src/utils/weakCache.ts [hunk 1]
```diff
@@ -38,3 +40,20 @@ export const setWeakCacheItem = <T>(
     deps = rest
   }
 }
+
+export const createMemoizeAtom = () => {
```
Comment (Thisen) [inline src/utils/weakCache.ts:44]: Nitpick: I feel like `cacheAtom` is better naming, but that's obviously subjective.
 
```suggestion
export const createCacheAtomFn = () => {
```

File: src/utils/weakCache.ts [hunk 2]
```diff
@@ -38,3 +40,20 @@ export const setWeakCacheItem = <T>(
...
+  const cache: WeakCache<Atom<unknown>> = new WeakMap()
+  const memoizeAtom = <AtomType extends Atom<unknown>, Deps extends object[]>(
```
Comment (Thisen) [inline src/utils/weakCache.ts:46]:
```suggestion
  const cacheAtom = <AtomType extends Atom<unknown>, Deps extends readonly object[]>(
```

File: src/utils/weakCache.ts [hunk 3]
```diff
@@ -38,3 +40,20 @@ export const setWeakCacheItem = <T>(
...
+  ) => {
+    const cachedAtom = getWeakCacheItem(cache, deps)
+    if (cachedAtom) {
+      return cachedAtom as AtomType
+    }
+    const createdAtom = createAtom()
+    setWeakCacheItem(cache, deps, createdAtom)
+    return createdAtom
+  }
+  return memoizeAtom
```
Comment (Thisen) [inline src/utils/weakCache.ts:58]:
```suggestion
  return cacheAtom
```

**Review by Thisen — APPROVED** [review]

**Comment by dai-shi** [comment]
Thanks. 

I don't have much preference on naming the function `memoizeAtom()` or `cacheAtom()`. Let's hear what others think. @aulneau @Aslemammad or anyone seeing this pr.

**Comment by Aslemammad** [comment]
It's remarkably subjective, but here, I think it's not just a simple caching; it's more about memoizing. Am I wrong? so I'd like to go with `memoizeAtom`

**Comment by aulneau** [comment]
I agree it feels more like memoization than caching to me. The distinction is I just want one instance of this atom, where if I wanted something cached, I'd assume the atom value would be involved. But again, very subject as others have said.

**Comment by dai-shi** [comment]
Thanks everyone. I will keep the current naming. It feels less confusing to me.

**Pull request merged**